pkg/api/v20231122/openshiftcluster_validatestatic.go

```diff
@@ -0,0 +1,404 @@
+// Licensed under the Apache License 2.0.
+
+import (
+	"fmt"
+	"net"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/Azure/go-autorest/autorest/azure"
+
+	"github.com/Azure/ARO-RP/pkg/api"
+	"github.com/Azure/ARO-RP/pkg/api/util/immutable"
+	apisubnet "github.com/Azure/ARO-RP/pkg/api/util/subnet"
+	"github.com/Azure/ARO-RP/pkg/api/validate"
+	"github.com/Azure/ARO-RP/pkg/util/pullsecret"
+	"github.com/Azure/ARO-RP/pkg/util/uuid"
+)
+
+type openShiftClusterStaticValidator struct {
+	location            string
+	domain              string
+	requireD2sV3Workers bool
+	resourceID          string
+
+	r azure.Resource
+}
+
+// Validate validates an OpenShift cluster
+func (sv openShiftClusterStaticValidator) Static(_oc interface{}, _current *api.OpenShiftCluster, location, domain string, requireD2sV3Workers bool, resourceID string) error {
+	sv.location = location
+	sv.domain = domain
+	sv.requireD2sV3Workers = requireD2sV3Workers
+	sv.resourceID = resourceID
+
+	oc := _oc.(*OpenShiftCluster)
+
+	var current *OpenShiftCluster
+	if _current != nil {
+		current = (&openShiftClusterConverter{}).ToExternal(_current).(*OpenShiftCluster)
+	}
+
+	var err error
+	sv.r, err = azure.ParseResourceID(sv.resourceID)
+	if err != nil {
+		return err
+	}
+
+	err = sv.validate(oc, current == nil)
+	if err != nil {
+		return err
+	}
+
+	if current == nil {
+		return nil
+	}
+
+	return sv.validateDelta(oc, current)
+}
+
+func (sv openShiftClusterStaticValidator) validate(oc *OpenShiftCluster, isCreate bool) error {
+	if !strings.EqualFold(oc.ID, sv.resourceID) {
+		return api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeMismatchingResourceID, "id", "The provided resource ID '%s' did not match the name in the Url '%s'.", oc.ID, sv.resourceID)
+	}
+	if !strings.EqualFold(oc.Name, sv.r.ResourceName) {
+		return api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeMismatchingResourceName, "name", "The provided resource name '%s' did not match the name in the Url '%s'.", oc.Name, sv.r.ResourceName)
+	}
+	if !strings.EqualFold(oc.Type, resourceProviderNamespace+"/"+resourceType) {
+		return api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeMismatchingResourceType, "type", "The provided resource type '%s' did not match the name in the Url '%s'.", oc.Type, resourceProviderNamespace+"/"+resourceType)
+	}
+	if !strings.EqualFold(oc.Location, sv.location) {
+		return api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeInvalidParameter, "location", "The provided location '%s' is invalid.", oc.Location)
+	}
+
+	return sv.validateProperties("properties", &oc.Properties, isCreate)
+}
+
+func (sv openShiftClusterStaticValidator) validateProperties(path string, p *OpenShiftClusterProperties, isCreate bool) error {
+	switch p.ProvisioningState {
+	case ProvisioningStateCreating, ProvisioningStateUpdating,
+		ProvisioningStateAdminUpdating, ProvisioningStateDeleting,
+		ProvisioningStateSucceeded, ProvisioningStateFailed:
```

Comment:
diff: missing `ProvisioningStateCancelled` that's in preview.

Reply:
Added, thanks.